control/diagnostics_tool/src/can_monitor.c

```diff
@@ -0,0 +1,335 @@
+        const int deltaT_message_arrival,
+        const int array_index )
+{
+    const float factor = 0.4;
+
+    int msg_freq = can_msg_array[ array_index ].msg_frequency;
+
+    int new_freq = deltaT_message_arrival;//1000 /deltaT_message_arrival;
+
+    return new_freq * factor + ( 1 - factor ) * msg_freq;
+}
+
+
+//
+static int calc_msg_timestamp_freq(
+        const int deltaT_message_arrival,
+        const int array_index )
+{
+    const float factor = 0.4;
+
+    int msg_freq = can_msg_array[ array_index ].msg_timestamp_frequency;
+
+    int new_freq = deltaT_message_arrival; //1000 / deltaT_message_arrival;
+
+    return new_freq * factor + ( 1 - factor ) * msg_freq;
+}
+
+
+//
+static void update_can_array_timestamp(
+        const int array_index,
+        const unsigned long tstamp )
+{
+    int deltaT_arrival =
+            get_timestamp() -
+            can_msg_array[ array_index ].last_arrival_timestamp;
+
+    can_msg_array[ array_index ].last_arrival_timestamp = get_timestamp();
+
+    int msg_timestamp_deltaT_arrival =
+            tstamp -
+            can_msg_array[ array_index ].frame_contents.tstamp;
+
+    can_msg_array[ array_index ].frame_contents.tstamp = tstamp;
+
+    can_msg_array[ array_index ].last_msg_deltaT = deltaT_arrival;
+
+    can_msg_array[ array_index ].last_msg_timestamp_deltaT =
+            msg_timestamp_deltaT_arrival;
+
+    can_msg_array[ array_index ].msg_frequency =
+            calc_msg_freq( deltaT_arrival, array_index );
+
+    can_msg_array[ array_index ].msg_timestamp_frequency =
+            calc_msg_timestamp_freq( msg_timestamp_deltaT_arrival, array_index );
+}
+
+
+
+
+// *****************************************************
+// public definitions
+// *****************************************************
+
+
+//
+unsigned long long get_timestamp()
+{
+    gettimeofday( &tv, NULL );
+
+    return ( unsigned long long )( tv.tv_sec ) * 1000 +
+            ( unsigned long long )( tv.tv_usec ) / 1000;
+}
+
+
+//
+void init_can_msg_array()
+{
+    int i;
+
+    for( i = 0; i < CAN_MSG_ARRAY_SIZE; i ++ )
+    {
+        can_msg_array[ i ].can_id = 0;
+    }
+}
+
+
+//
+const can_frame_s * const get_can_msg_array_index_reference(
+        const long can_id )
+{
+    const can_frame_s * frame_pointer;
+
+    int array_index = get_can_msg_array_index( can_id );
+
+    if( array_index == -1 )
+    {
+        frame_pointer = NULL;
+    }
+    else
+    {
+        frame_pointer = &can_msg_array[ array_index ];
+    }
+
+    return frame_pointer;
+}
+
+
+//
+void print_can_array( int * can_id_print_list, int num_can_ids )
+{
+    int i, j;
+
+    int print_all_ids = 0;
+
+    char print_array[ 500 ];
+
+    if( num_can_ids == 0 )
+    {
+        print_all_ids = 1;
+    }
+
+    add_line( "" );
+    add_line( "Raw CAN dump:" );
+
+    for( i = 0; i < CAN_MSG_ARRAY_SIZE; i++ )
+    {
+        if( print_all_ids && can_msg_array[ i ].can_id > 0 )
+        {
+            sprintf( print_array,
```

Comment:
If the sprintf calls are all the same, it might be useful to create a helper function in the case where you want to change what they say.